test/cryptonode_handlers_test.cpp

```diff
@@ -0,0 +1,243 @@
+#include <gtest/gtest.h>
+#include <inout.h>
+#include <jsonrpc.h>
+#include <graft_manager.h>
+#include <router.h>
+#include <requests.h>
+#include <requests/getinforequest.h>
+#include <requests/sendrawtxrequest.h>
+
+
+
+// cryptonode includes
+#include <wallet/wallet2.h>
+
+
+#include <string>
+#include <thread>
+#include <chrono>
+
+
+
+using namespace graft;
+using namespace std::chrono_literals;
+
+struct CryptonodeHandlersTest : public ::testing::Test
+{
+
+
+    void startServer()
+    {
+        ServerOpts sopts {"localhost:8855", "localhost:8856", 5.0, 4, 4, "localhost:28881/sendrawtransaction"};
+        Router router;
+        graft::registerGetInfoRequest(router);
+        graft::registerSendRawTxRequest(router);
+
+        Manager manager(sopts);
+        manager.addRouter(router);
+        manager.enableRouting();
+        this->manager = &manager;
+        server.serve(manager.get_mg_mgr());
+    }
+
+    void stopServer()
+    {
+
+    }
+
+    CryptonodeHandlersTest()
+    {
+        mlog_configure("", true);
+        mlog_set_log_level(2);
+        LOG_PRINT_L0("L0");
+        LOG_PRINT_L1("L1");
+        LOG_PRINT_L2("L2");
+
+        server_thread = std::thread([this]() {
+            this->startServer();
+        });
+
+        LOG_PRINT_L0("Server thread started..");
+
+        while (!server.ready) {
+            LOG_PRINT_L0("waiting for server");
+            std::this_thread::sleep_for(1s);
+        }
+
+        LOG_PRINT_L0("Server ready");
+    }
+
+    ~CryptonodeHandlersTest()
+    {
+    }
+
+
+    static std::string run_cmdline_read(const std::string& cmdl)
+    {
+        FILE* fp = popen(cmdl.c_str(), "r");
+        assert(fp);
+        std::string res;
+        char path[1035];
+        while(fgets(path, sizeof(path)-1, fp))
+        {
+            res += path;
+        }
+        pclose(fp);
+        return res;
+    }
+    static std::string escape_string_curl(const std::string & in)
+    {
+        std::string result;
+        for (char c : in) {
+            if (c == '"')
+                result.push_back('\\');
+            result.push_back(c);
+        }
+        return result;
+    }
+
+    static std::string send_request(const std::string &url, const std::string &json_data)
+    {
+        std::ostringstream s;
+        s << "curl -s -H \"Content-type: application/json\" --data \"" << json_data << "\" " << url;
+        std::string ss = s.str();
+        return run_cmdline_read(ss.c_str());
+    }
+
+    static std::string send_request(const std::string &url)
+    {
+        std::ostringstream s;
+        s << "curl -s " << url;
+        std::string ss = s.str();
+        return run_cmdline_read(ss.c_str());
+    }
+
+
+    static void SetUpTestCase()
+    {
+    }
+
+    static void TearDownTestCase()
+    {
+    }
+
+    virtual void SetUp() override
+    { }
+    virtual void TearDown() override
+    { }
+
+    GraftServer server;
+    Manager   * manager;
+    std::thread server_thread;
```

Comment:
Why not?
Manager  manager; //not a pointer
GraftServer server;
Please, pay attention to the order, in this case server dtor will be called before manager call. And you can place server_thread.join into ~CryptonodeHandlersTest().
Also note, according to the mongoose, mg_mgr_free should close all open connections, including listening.
Also, as I know CryptonodeHandlersTest object are created for each TEST_F, and the order of TEST_Fs is not specified.

Reply:
> And you can place server_thread.join into ~CryptonodeHandlersTest().
intent was **explicitly** wait for server thread in the test code